### PR TITLE
summarize_networks InvalidIndexError

### PR DESCRIPTION
**Pull request by ntalluri**
PR for https://github.com/Reed-CompBio/spras/issues/188

Currently created a test case to showcase the error and how it is happening

### CODE REVIEW

File: test/ml/test_ml.py [hunk 1]
```diff
@@ -42,6 +42,9 @@ def test_summarize_networks_wrong_direction(self):
         with pytest.raises(ValueError):
             ml.summarize_networks([INPUT_DIR + 'test-data-wrong-direction/wrong-direction.txt'])
 
+    def test_repeat_edges(self):
```
Comment (ntalluri) [inline test/ml/test_ml.py:45]: Look at `run tests` in checks. The error is the expected behavior I was expecting to see.

File: spras/analysis/ml.py [hunk 2]
```diff
@@ -78,6 +78,7 @@ def summarize_networks(file_paths: Iterable[Union[str, PathLike]]) -> pd.DataFra
                 str(tup[0]): 1,
             }, index=tup[1]
         )
+        dataframe = dataframe[~dataframe.index.duplicated(keep='first')]
```
Comment (ntalluri) [inline spras/analysis/ml.py:81]: https://pandas.pydata.org/docs/reference/api/pandas.Index.duplicated.html

Comment (ntalluri) [inline spras/analysis/ml.py:81]: add a warning that edges are being dropped from a file

Comment (agitter) [reply spras/analysis/ml.py:81]:
```suggestion
        # Mark the first occurrence of a duplicate as True and all others as False
        # Non-duplicates are also marked as True
        dataframe = dataframe[~dataframe.index.duplicated(keep='first')]
```

Comment (agitter) [reply spras/analysis/ml.py:81]: If this is hard to do here, we can think about whether it should be handled somewhere else. For instance, should the `parse_output` functions check for duplicate edges and raise a warning or error?

**Review by agitter — COMMENTED** [review]
What you have looks good.

If you didn't already open an issue about MinCostFlow giving the duplicate issues that trigged this problem, please do that and link it to https://github.com/Reed-CompBio/spras/issues/188#issuecomment-2438619614. We may want to follow up on that potential bug later.

File: test/parse-outputs/input/duplicate-edges/omicsintegrator1-raw-pathway.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+A	pd	C
+C	pp	D
+A	pd	C
+C	pp	D
```
Comment (agitter) [inline test/parse-outputs/input/duplicate-edges/omicsintegrator1-raw-pathway.txt:4]: This test case makes it seem like the util function may be too simplistic. What if the edges were instead:
```
A	pd	C
C	pp	D
A	pp	C
D	pp	C
```
What would the current behavior be? What would the desired behavior be?

File: spras/util.py [hunk 4]
```diff
@@ -76,3 +76,14 @@ def raw_pathway_df(raw_pathway_file: str, sep: str = '\t', header: int = None) -
         df = pd.DataFrame(columns=['Node1', 'Node2', 'Rank', 'Direction'])
 
     return df
+
+def duplicate_edges(df: pd.DataFrame):
```
Comment (agitter) [inline spras/util.py:81]: I like this new design of checking for duplicate edges in parse outputs. However, see the comment on the Omics Integrator test case about whether we need to have more complex duplicate checking logic.

**Review by agitter — COMMENTED** [review]
I pushed a note to the contributing guide about this new post-processing step.

File: test/parse-outputs/input/duplicate-edges/omicsintegrator1-raw-pathway.txt (hunk 3 above)
Comment (ntalluri) [reply test/parse-outputs/input/duplicate-edges/omicsintegrator1-raw-pathway.txt:4]: With that, the code in util.py doesn't take that my thoughts into consideration. It will remove anything in the form A C without taking into account the directionality.

File: spras/util.py (hunk 4 above)
Comment (ntalluri) [reply spras/util.py:81]: I updated it to remove duplicate rows now; treats undirected and directed edges coming between 2 nodes as separate edges that the original algorithm should have dealt with.

**Comment by ntalluri** [comment]
> I pushed a note to the contributing guide about this new post-processing step.
> 
> > If you didn't already open an issue about MinCostFlow giving the duplicate issues that trigged this problem, please do that and link it to #188 (comment). We may want to follow up on that potential bug later.
> 
> Do we still want a new issue to track this?

I think we should track that there is still a problem with MCF, but we solve the problem for now with this PR.

**Comment by ntalluri** [comment]
keep the smallest rank instead of keep first; sort by rank first
need to sort the undirected edges ( B - A changed to A - B)  before removing duplicates

**Comment by ntalluri** [comment]
@agitter Ready for another review

File: test/parse-outputs/input/duplicate-edges/omicsintegrator1-raw-pathway.txt [hunk 5]
```diff
@@ -2,3 +2,6 @@ A	pd	C
 C	pp	D
 A	pd	C
 C	pp	D
+A	pp	C
```
Comment (agitter) [inline test/parse-outputs/input/duplicate-edges/omicsintegrator1-raw-pathway.txt:5]: This is a good test case for all of the corner cases we want to cover.

**Review by agitter — APPROVED** [review]
